### PR TITLE
[apache/helix] -- Updated code to be build compatible with JDK-8.

### PR DESCRIPTION
**Pull request by himanshukandwal**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:
Updated codebase to be build with JDK-8.

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
We will releasing the Open source Helix on Java 11, but we are updating the codebase to be build compatible with Java-8, so that we can internally release it on Java-8.

### Tests

- [x] The following tests are written for this issue:

```console
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Helix 1.3.2-SNAPSHOT:
[INFO] 
[INFO] Apache Helix ....................................... SUCCESS [  0.981 s]
[INFO] Apache Helix :: Metrics Common ..................... SUCCESS [  1.187 s]
[INFO] Apache Helix :: Metadata Store Directory Common .... SUCCESS [  0.618 s]
[INFO] Apache Helix :: ZooKeeper API ...................... SUCCESS [  1.571 s]
[INFO] Apache Helix :: Helix Common ....................... SUCCESS [  0.498 s]
[INFO] Apache Helix :: Core ............................... SUCCESS [  7.787 s]
[INFO] Apache Helix :: Admin Webapp ....................... SUCCESS [  0.994 s]
[INFO] Apache Helix :: Restful Interface .................. SUCCESS [  1.910 s]
[INFO] Apache Helix :: Distributed Lock ................... SUCCESS [  0.375 s]
[INFO] Apache Helix :: HelixAgent ......................... SUCCESS [  0.595 s]
[INFO] Apache Helix :: Front End .......................... SUCCESS [01:04 min]
[INFO] Apache Helix :: Recipes ............................ SUCCESS [  0.013 s]
[INFO] Apache Helix :: Recipes :: Rabbitmq Consumer Group . SUCCESS [  0.497 s]
[INFO] Apache Helix :: Recipes :: Rsync Replicated File Store SUCCESS [  0.519 s]
[INFO] Apache Helix :: Recipes :: distributed lock manager  SUCCESS [  0.406 s]
[INFO] Apache Helix :: Recipes :: distributed task execution SUCCESS [  0.392 s]
[INFO] Apache Helix :: Recipes :: service discovery ....... SUCCESS [  0.407 s]
[INFO] Apache Helix :: View Aggregator .................... SUCCESS [  0.697 s]
[INFO] Apache Helix :: Meta Client ........................ SUCCESS [  0.568 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:24 min
[INFO] Finished at: 2023-10-09T15:21:35-07:00
[INFO] ------------------------------------------------------------------------
```

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java [hunk 1]
```diff
@@ -100,7 +100,7 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
           getNodeWithHighestPoints(replica, nodes, clusterModel.getContext(), busyInstances,
               optimalAssignment);
       // stop immediately if any replica cannot find best assignable node
-      if (maybeBestNode.isEmpty() || optimalAssignment.hasAnyFailure()) {
+      if (!maybeBestNode.isPresent() || optimalAssignment.hasAnyFailure()) {
```
Comment (desaikomal) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java:103]: any particular reason to change this particular check? this is optional property.

**Review by desaikomal — APPROVED** [review]
Thanks @himanshukandwal - looks good to me.

**Comment by himanshukandwal** [comment]
This PR is ready to be merged.

Final Commit Message:
We will releasing the Open source Helix on Java 11, but we are updating the codebase to be build compatible with Java-8, so that we can internally release it on Java-8.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java (hunk 1 above)
Comment (desaikomal) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java:103]: thanks.. i thought the data structure is the same, but here method is the issue. thanks.

**Review by xyuanlu — APPROVED** [review]

**Pull request merged**